server/src/main/java/com/linecorp/centraldogma/server/command/AbstractCommandExecutor.java

```diff
@@ -97,8 +102,24 @@ public final void setWritable(boolean writable) {
     @Override
     public final <T> CompletableFuture<T> execute(Command<T> command) {
         requireNonNull(command, "command");
-        if (!isWritable()) {
-            throw new IllegalStateException("running in read-only mode");
+        if (!isWritable() && !(command instanceof AdministrativeCommand)) {
+            // Reject all commands except for AdministrativeCommand when the replica is in read-only mode.
+            // AdministrativeCommand is allowed because it is used to change the read-only mode or migrate
+            // metadata under maintenance mode.
+            throw new IllegalStateException("running in read-only mode. command: " + command);
+        }
+
+        if (command.type() == CommandType.UPDATE_SERVER_STATUS) {
```

Comment:
It was added to stop `ZooKeeperCommandExecutor` but it does not look good. I will move the logic to `ZooKeeperCommandExecutor.blockingExecute()`.

Reply:
>  It was added to stop ZooKeeperCommandExecutor but it does not look good. I will move the logic to ZooKeeperCommandExecutor.blockingExecute().

May I ask why is this needed?

The executing replica will execute `delegate.execute` here:

https://github.com/line/centraldogma/blob/b82680e0a7bdc9fe8dfe754de755fcbb76f8b351/server/src/main/java/com/linecorp/centraldogma/server/internal/replication/ZooKeeperCommandExecutor.java#L1086

And the other replicas will execute `delegate.execute` here:
https://github.com/line/centraldogma/blob/b82680e0a7bdc9fe8dfe754de755fcbb76f8b351/server/src/main/java/com/linecorp/centraldogma/server/internal/replication/ZooKeeperCommandExecutor.java#L727

Both will reach here:

https://github.com/line/centraldogma/blob/b82680e0a7bdc9fe8dfe754de755fcbb76f8b351/server/src/main/java/com/linecorp/centraldogma/server/command/StandaloneCommandExecutor.java#L191

am I missing anything? 😅

Reply:
Unlike other commands that are delegated to `StandaloneCommandExecutor`, `UpdateServerStatusCommand` should be applied to each `CommandExecutor` to stop both executors.

Reply:
I see, I didn't realize `writable` is set per-executor 😅 

This behavior makes me realize that if we issue a `UpdateServerStatusCommand` with `writable = false`, then we can't recover a replica with the `AdministrativeService` API (since it will only set `writable` to the outermost executor). That is, we can only recover it via issuing another `UpdateServerStatusCommand`.

Reply:
Right. The behavior is different. `AdministrativeService.updateStatus()` applies `writable` and `replicating` to the server only. Although `writable` recovers the server's status but can't update the whole cluster.

As said, we need to integrate `AdministrativeService` with `UpdateServerStatusCommand`. https://github.com/line/centraldogma/pull/899#issuecomment-1916634886